machine-learning/02-openpbta_consensus_clustering.Rmd

```diff
@@ -0,0 +1,378 @@
+  select(gene_id, all_of(medulloblastoma_samples)) %>%
+  tibble::column_to_rownames("gene_id") %>%
+  as.matrix()
+```
+
+## Cluster validation
+
+### Consensus clustering
+
+The first method we'll use is called consensus clustering.
+Consensus clustering to finds the "consensus" across multiple runs of the algorithm using a resampling procedure.
+
+We'll use the package [`ConsensusClusterPlus`](https://bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html) that we loaded up top.
+
+The consensus clustering methodology was first introduced in [Monti et al. _Machine Learning_. 2003.](https://doi.org/10.1023/A:1023949509487)
+
+Consensus clustering is one way to help you determine the number of clusters in your data, but it is not the only methodology available. 
+Check out the [Data Novia course _Cluster Validation Essentials_ by Alboukadel Kassambara](https://www.datanovia.com/en/courses/cluster-validation-essentials/) for a deeper dive.
+
+```{r consensus_clustering}
+cc_results <- ConsensusClusterPlus(rnaseq_mat,
+                                   maxK = 15,
+                                   # Setting this seed is necessary for the 
+                                   # results to be reproducible
+                                   seed = 2020,
+                                   innerLinkage = "average",
+                                   finalLinkage = "average",
+                                   distance = "pearson")
+```
+
+It seems like there are 3 main stable clusters from the consensus clustering results.
+A cluster of a few samples may not be that helpful in reaching our analysis goals.
+
+Let's take a look at the class labels for _k_ = 9.
+
+```{r cc_labels, live = TRUE}
+# table() creates a contingency table of counts
+# 
+table(cc_results[[9]]$consensusClass)
+```
+
+(Note: the numbering of the clusters is arbitrary here.)
+
+**Let's consider clusters 1, 3, and 4 for further analysis.**
+But first, we'll save the entirety of the consensus clustering results to file.
+
+```{r save_cc_results, live = TRUE}
+# Write consensus clustering results to file
+write_rds(cc_results, path = cc_results_file)
+```
+
+And now to extract the samples in the clusters of interest.
+What we used `table()` on before is actually a named vector.
+
+```{r extract_cluster_labels}
+cc_cluster_labels <- cc_results[[9]]$consensusClass
+head(cc_cluster_labels)
+```
+
+We'll extract the names (biospecimen IDs) for samples in clusters 1, 3, and 4, which contain the majority of samples.
+
+```{r sample_names}
+sample_index <- which(cc_cluster_labels %in% c(1, 3, 4))
+samples_in_clusters <- names(cc_cluster_labels)[sample_index]
+```
+
+Check what proportion of total samples are in one of these three clusters.
+
+```{r proportion_in_clusters}
+# number of samples that met our logical criterion above divided by total 
+# number of samples
+length(sample_index) / length(cc_cluster_labels)
+```
+
+And now we'll filter the RNA-seq matrix.
```

Comment:
Are we choosing to ignore samples that don't cluster? If so, we should justify this?

Reply:
Only for the silhouette value calculations because doing that for a cluster of 1 sample (or 2, for that matter) doesn't make a lot of sense to me. I'm also happy to rip out the silhouette value stuff to save on time to be honest. (I feel like this is a bit of an odd use without enough context.) The main reason for including consensus clustering is _skepticism_ about clustering results.

Thoughts?

Reply:
That makes sense. 

I do wonder if the silhouette juice is worth the squeeze. If you spend more time on the "consensus because clusters can be janky" I think that makes the rhetorical point.

Reply:
:chopped: